### DISCUSSION TITLE
where do I find the API documentation for the tulip, amy, midi python libraries?

### DISCUSSION THREAD

**Discussion by ely-peddler** [shorepine/tulipcc#393] [General]
There's lots of great example code in the docs here but I can't find the API documentation covering the classes and methods within the tulip, amy & midi packages. 
For example what are the method signatures for the synth object used in the music tutorial?

**Comment by bwhitman** [comment]
Nothing yet, that's a todo https://github.com/shorepine/tulipcc/issues/194

You'll have to look at the code itself for now for everything, you'll find the python modules that ship with Tulip here: https://github.com/shorepine/tulipcc/tree/main/tulip/shared/py

**Reply by ely-peddler** [reply]
ah ok thanks - perhaps just add a link to that on the API page for now?

**Comment by ely-peddler** [comment]
I've had a dig around and there's something I still can't grasp

How do these 2 bits of code work together?

def note(t):
    synth.note_on(random.choice(chord), 0.6, time=t)

slot = tulip.seq_add_callback(note, 48)

Is the value 48 passed to note(t) as the variable t?
Is note(t) called every 48 ticks? Or does note_on(...time=t) set the rate?

**Reply by bwhitman** [reply]
See more here: https://github.com/shorepine/tulipcc/blob/main/docs/tulip_api.md#music-sequencer

`tulip.seq_add_callback` takes 2 parameters, one is a python function that gets called by the sequencer, and the other is a `divider` on the system sequencer. The system sequencer is called by default 48 times a quarter note (aka `48 ppq`). Most things you'll want to do on Tulip don't need that sort of resolution, so for most people, you'll want to use a high divider, like 48. Setting the divider to 48 means your function will be called 48/**48** times a quarter note, so just once a quarter note. If you set it to 24, you'll be called twice a quarter note.

the `t` parameter sent to the callback `note` is for the absolute `time` (in milliseconds) of the event and is sent by the sequencer. The idea is that the sequencer callback will trigger slightly _before_ you need it to, because often Python code is slower and you want to schedule AMY synth events at the correct time. So you pass the `t` to AMY. As long as you want things to stay "in time" you shouldn't need to worry about `t` -- just pass it along to any direct AMY or `Synth` messages you're sending.

**Reply by ely-peddler** [reply]
Oh I see. 

```
slot = tulip.seq_add_callback(note, 48) # calls note 'just before' every 1/4 note, if it was 24 it would be just before every 1/8 note

def note(t): # t is the exact time the 1/4 (or 1/8) note triggers
  synth.note_on(random.choice(chord), 0.6, time=t) # tells the syth to trigger the note_on exactly on the 1/4 (or 1/8) note
```
